tokenizers/src/tokenizer/mod.rs

```diff
@@ -795,22 +800,37 @@ where
     }
 
     /// Decode the given ids, back to a String
-    pub fn decode(&self, ids: &[u32], skip_special_tokens: bool) -> Result<String> {
-        let tokens = ids
-            .iter()
-            .filter_map(|id| {
-                self.added_vocabulary
-                    .id_to_token(*id, &self.model)
-                    .filter(|token| {
-                        !skip_special_tokens || !self.added_vocabulary.is_special_token(token)
-                    })
-            })
-            .collect::<Vec<_>>();
-
+    pub fn decode(
+        &self,
+        ids: &[u32],
+        skip_special_tokens: bool,
+        spaces_between_added_tokens: bool,
+    ) -> Result<String> {
+        // split on added_tokens
+
+        let mut tokens_to_decode: Vec<String> = Vec::new();
+        let last_idx = ids.len();
+        let push_spaces = !&self.decoder.is_some();
+
+        for (index, id) in ids.iter().enumerate() {
+            if let Some(token) = self.added_vocabulary.id_to_token(*id, &self.model) {
+                if self.added_vocabulary.is_special_token(&token) && skip_special_tokens {
+                    continue;
+                } else if (self.added_vocabulary.get_added_tokens().contains(&token)
+                    && spaces_between_added_tokens
+                    || push_spaces)
+                    && index < last_idx - 1
+                {
+                    tokens_to_decode.extend([token, " ".to_string()]);
+                } else {
+                    tokens_to_decode.push(token);
+                }
```

Comment:
I'm not sure this works.

This is already tons better than before, but you're still modifying "tokens_to_decode" by introducing extra spaces.

The `push_spaces` is really convoluted (it's defined as double negative) and the you use `||` in the condition, meaning we always add if there's no decoder, but we add only if it's an added token when there's a decoder.

Refactoring the conditions names might make this a bit more obvious what we're trying to achieve.

Also adding extra "tokens" containing `" "` will *definitely* have unintended consequences in the decoders....
I think this kind of invalidates the approach.

Another idea would be to "add a decoder" at the end of the chain that would deal with those added tokens or somethign like that. In any case, "faking" tokens will yield issues I think.

Reply:
Refactored it!